### PR TITLE
feat(tv): add code to compute total variation

### LINKED ISSUE #17: Implement Total variation metric/loss
**Is your feature request related to a problem? Please describe.**
It's been used in a couple of Style Transfer articles. 
https://www.wikiwand.com/en/Total_variation_denoising

**Describe the solution you'd like**
Implement anisotropic version of total variation.

### PR DESCRIPTION
**Pull request by zakajd**
Finished  Total Variation metric
Closes #17 

Part of the tests failed due to torchvision import error. That looks like CI problem.

### CODE REVIEW

File: photosynthesis_metrics/ssim.py [hunk 1]
```diff
@@ -14,6 +14,8 @@
 from torch.nn.modules.loss import _Loss
 from typing import Union, Optional, List, Tuple
 
+from .utils import _adjust_dimensions, _validate_input
```
Comment (snk4tr) [inline photosynthesis_metrics/ssim.py]: Why do you prefer relative imports to absolute ones?

File: photosynthesis_metrics/tv.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        h_variance = torch.sum(torch.pow(x[:,:,1:,:] - x[:,:,:-1,:], 2), dim=[1, 2, 3])  
+        tv_val = (h_variance + w_variance)
+    else:
+        raise ValueError("Incorrect reduction type, should be one of {'l1', 'l2', 'l2_squared'}")
+
+    if size_average:
+        return tv_val.mean()
+    else:
+        return tv_val
+
```
Comment (snk4tr) [inline photosynthesis_metrics/tv.py]: Formatting: add extra blank line here. 

From the Google style guide (section 3.5):
**Two blank lines** between top-level definitions, be they function or class definitions. **One blank line** between method definitions and between the class line and the first method. No blank line following a def line. Use single blank lines as you judge appropriate within functions or methods.

File: photosynthesis_metrics/tv.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+def total_variation(x: torch.Tensor, size_average: bool = True, reduction_type: str = 'l2') -> torch.Tensor:
+    r"""Interface of Total Variation.
+
+    Args:
+        x: Batch of images. Required to be 4D, channels first (N,C,H,W).
+        size_average: If size_average=True, total_variation of all images will be averaged as a scalar.
+        reduction_type: one of {'l1', 'l2', 'l2_squared'}
+
+    Returns:
+        Value of Structural Similarity (SSIM) index.
```
Comment (snk4tr) [inline photosynthesis_metrics/tv.py]: Docs: I hope you compute not SSIM, but TV here :)

File: photosynthesis_metrics/tv.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    References:
+        .. [1] Wang, Z., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P.
+           (2004). Image quality assessment: From error visibility to
+           structural similarity. IEEE Transactions on Image Processing,
+           13, 600-612.
+           https://ece.uwaterloo.ca/~z70wang/publications/ssim.pdf,
+           :DOI:`10.1109/TIP.2003.819861`
```
Comment (snk4tr) [inline photosynthesis_metrics/tv.py]: Docs: the reference also corresponds to SSIM. I believe it is not needed here.

File: photosynthesis_metrics/tv.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    """
+    _validate_input(x=x, y=x, kernel_size=3)
```
Comment (snk4tr) [inline photosynthesis_metrics/tv.py]: I like your idea to reuse as much code as possible. Please consider removing `kernel_size=3` part since `kernal_size=None` by default and this case is handled within the function.

File: photosynthesis_metrics/utils.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    elif num_dimentions == 3:
+        x = x.expand(1, *x.shape)
+        y = y.expand(1, *y.shape)
+    elif num_dimentions != 4:
+        raise ValueError('Expected 2, 3, or 4 dimensions (got {})'.format(num_dimentions))
+
+    return x, y
+
+
+def _validate_input(x: torch.Tensor, y: torch.Tensor, kernel_size: Optional[int] = None,
```
Comment (snk4tr) [inline photosynthesis_metrics/utils.py]: Please consider making `y: Optional[torch.Tensor] = None`. It will remove the need for you to do things like `y=x` in your total variance interface.

File: photosynthesis_metrics/tv.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    Args:
+        x: Tensor of shape :math:`(N, C, H, W)` holding an input image.
+        size_average: If size_average=True, total variation of all images will be averaged as a scalar.
+        reduction_type: {'l1', 'l2', 'l2_squared'}, defines which type of norm to implement, isotropic  or anisotropic.
+
+    Returns:
+        tv : Total variation of a given tensor
+    """
+
+    assert isinstance(x, torch.Tensor), f'Image stack must be of type torch.Tensor, got {type(x)} instead.'
```
Comment (snk4tr) [inline photosynthesis_metrics/tv.py]: Why did you decide to do this assertion here? 
If this function is called from the interface (`def total_variation`), then this and other requirements have to be checked in the interface (that is its purpose).
If this function may be called externally, it should not be protected (single underscore before function name).

File: photosynthesis_metrics/tv.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+        >>> loss = SSIMLoss()
+        >>> prediction = torch.rand(3, 3, 256, 256, requires_grad=True)
+        >>> target = torch.rand(3, 3, 256, 256)
+        >>> output = loss(prediction, target, max_val=1.)
+        >>> output.backward()
+    """
+
+    def __init__(self, size_average: bool = True, reduction_type: str = 'l2', reduction: str = 'mean'):
+
```
Comment (snk4tr) [inline photosynthesis_metrics/tv.py]: Formatting: remove blank line here.

From the Google style guide (section 3.5):
Two blank lines between top-level definitions, be they function or class definitions. One blank line between method definitions and between the class line and the first method. **No blank line following a def line**. Use single blank lines as you judge appropriate within functions or methods.

File: photosynthesis_metrics/tv.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        super(TVLoss, self).__init__()
+
+        # Generic loss parameters.
+        self.size_average = size_average
+        self.reduction_type = reduction_type
+        self.reduction = reduction
+
+        # Loss-specific parameters.
+        self.tv_func = functools.partial(_compute_tv, size_average=size_average, reduction_type=reduction_type)
```
Comment (snk4tr) [inline photosynthesis_metrics/tv.py]: Like it a lot!

File: photosynthesis_metrics/tv.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+        >>> output.backward()
+    """
+
+    def __init__(self, size_average: bool = True, reduction_type: str = 'l2', reduction: str = 'mean'):
+
+        super(TVLoss, self).__init__()
+
+        # Generic loss parameters.
+        self.size_average = size_average
+        self.reduction_type = reduction_type
```
Comment (snk4tr) [inline photosynthesis_metrics/tv.py]: Since you use partial below, you do not need to introduce these variables. They are never used.

File: tests/test_tv.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    return torch.rand(4, 3, 256, 256)
+
+
+# ================== Test function: `total_variation` ==================
+def test_tv_works(prediction: torch.Tensor) -> None:
+    measure = total_variation(prediction, size_average=True, reduction_type='l2')
+    assert measure > 0 
+    measure = total_variation(prediction, size_average=True, reduction_type='l1')
+    measure = total_variation(prediction, size_average=True, reduction_type='l2_squared')
+
```
Comment (snk4tr) [inline tests/test_tv.py]: Formatting: remove blank line here.

From the Google style guide (section 3.5):
**Two blank lines between top-level definitions**, be they function or class definitions. One blank line between method definitions and between the class line and the first method. No blank line following a def line. Use single blank lines as you judge appropriate within functions or methods.

File: tests/test_tv.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+
+# ================== Test class: `TVLoss` ==================
+def test_tvloss_init(prediction: torch.Tensor, target: torch.Tensor) -> None:
+    l = TVLoss()
+
```
Comment (snk4tr) [inline tests/test_tv.py]: Formatting: and extra blank line here.

From the Google style guide (section 3.5):
**Two blank lines between top-level definitions**, be they function or class definitions. One blank line between method definitions and between the class line and the first method. No blank line following a def line. Use single blank lines as you judge appropriate within functions or methods.

File: photosynthesis_metrics/tv.py (hunk 7 above)
Comment (zakajd) [reply photosynthesis_metrics/tv.py]: This is protected function, assert here is not necessary. Removed it
